extensions/amp-date-countdown/0.1/amp-date-countdown.js

```diff
@@ -0,0 +1,254 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ActionTrust} from '../../../src/action-trust';
+import {AmpEvents} from '../../../src/amp-events';
+import {Services} from '../../../src/services';
+import {createCustomEvent} from '../../../src/event-helper';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {removeChildren} from '../../../src/dom';
+import {user} from '../../../src/log';
+
+/** @const {string} */
+const TAG = 'amp-date-countdown';
+
+export class AmpDateCountdown extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @const {function(!Element)} */
+    this.boundRendered_ = this.rendered_.bind(this);
+
+    /** @private {string} */
+    this.endDate_ = '';
+
+    /** @private {number} */
+    this.timestampMs_ = 0;
+
+    /** @private {number} */
+    this.timestampSeconds_ = 0;
+
+    /** @private {number} */
+    this.offsetSeconds_ = 0;
+
+    /** @private {string} */
+    this.biggestUnit_ = '';
+
+    /** @const {!../../../src/service/template-impl.Templates} */
+    this.templates_ = Services.templatesFor(this.win);
+  }
+
+  /** @override */
+  buildCallback() {
+    //Note: One of datetime, timestamp-ms, timestamp-seconds is required.
+    this.endDate_ = this.element.getAttribute('end-date');
+    this.timestampMs_ = Number(this.element.getAttribute('timestamp-ms'));
+    this.timestampSeconds_ = Number(this.element.getAttribute('timestamp-seconds'));
+    this.offsetSeconds_ = Number(this.element.getAttribute('offset-seconds')) || 0;
+
+    const locale = this.element.getAttribute('locale') || 'en';
+    const whenEnded = this.element.getAttribute('when-ended') || 'stop';
+    this.biggestUnit_ = (this.element.getAttribute('biggest-unit')) || 'days';
+
+    const epoch = this.getEpoch_() + (this.offsetSeconds_ * 1000);
+    let differentBetween = new Date(epoch) - new Date();
+    let data;
+    const localeWordList = this.getLocaleWord_(locale);
+    const delay = 1000;
+
+    const countDownTicker = this.win.setInterval(() => {
```

Comment:
I think you should override `AMP.BaseElement#layoutCallback` and create the timer in that function, since the countdown doesn't need to update if the user hasn't scrolled down to the countdown yet.

Then override `AMP.BaseElement#unlayoutCallback` and call `this.win.clearInterval` there, to free up the timer resources.

Reply:
I did try and write the code based on your suggestion, not too sure am i writing it correctly, please review the code and advise me if it is not right.